cnap/core/metrics.py

```diff
@@ -0,0 +1,182 @@
+"""A Metrics module.
+
+This module provides a class to manage metrics in a Prometheus-based application.
+It provides methods to create, track, and update metrics.
+
+Classes:
+    MetricType: An enum for the type of metric.
+    ErrorMessage: An enum for error messages.
+    MetricsManager: A class to manage metrics in a Prometheus-based application.
+"""
+
+from enum import Enum
+from functools import wraps
+from typing import Any, Callable, Dict, Union
+from threading import Lock
+from prometheus_client import Counter, Gauge, Histogram, Summary
+
+class MetricType(Enum):
+    """An enum for the type of metric.
+
+    This enum defines the types of metrics that can be created.
+    """
+    COUNTER = 'Counter'
+    GAUGE = 'Gauge'
+    HISTOGRAM = 'Histogram'
+    SUMMARY = 'Summary'
+
+
+class ErrorMessage(Enum):
+    """An enum for error messages.
+
+    This enum defines the error messages that can be raised.
+    """
+    METRIC_EXISTS = 'Metric {} already exists'
+    INVALID_METRIC_TYPE = 'Invalid metric type: {}'
+    METRIC_DOES_NOT_EXIST = 'Metric {} does not exist'
+    METRIC_TYPE_MISMATCH = 'Metric {} is not a {}'
+
+
+class MetricsManager:
```

Comment:
Does this class need to use the singleton pattern to ensure that there is only one instance in each microservice?

Reply:
It makes sense